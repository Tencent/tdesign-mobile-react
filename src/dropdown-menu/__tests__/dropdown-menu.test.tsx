import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, fireEvent, act, within } from '@testing-library/react';
import { DropdownMenu, DropdownItem, DropdownOption } from '../index';

const productOptions: DropdownOption[] = [
  { value: 'all', label: '全部产品', disabled: false },
  { value: 'new', label: '最新产品', disabled: false },
  { value: 'hot', label: '最火产品', disabled: false },
  { value: 'disabled', label: '禁用选项', disabled: true },
];

describe('DropdownMenu', () => {
  describe('props', () => {
    it(':style', () => {
      const { container } = render(
        <DropdownMenu className="my-menu" style={{ color: 'rgb(255, 0, 0)' }}>
          <DropdownItem label="样式测试" options={productOptions} />
        </DropdownMenu>,
      );
      const root = container.querySelector('.my-menu') as HTMLElement;
      expect(Boolean(root)).toBe(true);
      expect(root.style.color === 'rgb(255, 0, 0)').toBe(true);
      expect(container).toMatchSnapshot();
    });

    it(':direction and zIndex', () => {
      const { container, getByText } = render(
        <DropdownMenu direction="up" zIndex={9999}>
          <DropdownItem label="产品" options={productOptions} />
        </DropdownMenu>,
      );

      expect(container.querySelector('.t-dropdown-menu')).toBeTruthy();

      fireEvent.click(getByText('产品'));
      const dropdownItem = container.querySelector('.t-dropdown-item') as HTMLElement;
      expect(dropdownItem).toBeTruthy();
      if (dropdownItem) {
        expect(dropdownItem.style.zIndex).toBe('9999');
      }
      expect(container).toMatchSnapshot();
    });

    it(':optionsColumns', () => {
      const { getByText } = render(
        <DropdownMenu>
          <DropdownItem label="双列" multiple optionsColumns={2} options={productOptions} />
        </DropdownMenu>,
      );
      fireEvent.click(getByText('双列'));
      expect(getByText('全部产品')).toBeTruthy();
      expect(document.body).toMatchSnapshot();
    });

    it(':footer', () => {
      const { getByText } = render(
        <DropdownMenu>
          <DropdownItem label="自定义" options={productOptions} footer={<div>自定义footer</div>} />
        </DropdownMenu>,
      );
      fireEvent.click(getByText('自定义'));
      expect(getByText('自定义footer')).toBeTruthy();
      expect(document.body).toMatchSnapshot();
    });

    it(':defaultValue', () => {
      const { getByText } = render(
        <DropdownMenu>
          <DropdownItem options={productOptions} defaultValue="new" />
        </DropdownMenu>,
      );
      expect(getByText('最新产品')).toBeTruthy();
      expect(document.body).toMatchSnapshot();
    });

    it(':keys', () => {
      const customOptions = [
        { k: 'a', l: 'A 选项', d: false },
        { k: 'b', l: 'B 选项', d: false },
      ] as unknown as DropdownOption[];
      const { getByText } = render(
        <DropdownMenu>
          <DropdownItem options={customOptions} keys={{ value: 'k', label: 'l', disabled: 'd' }} defaultValue="b" />
        </DropdownMenu>,
      );
      expect(getByText('B 选项')).toBeTruthy();
      expect(document.body).toMatchSnapshot();
    });
  });

  describe('event', () => {
    it(':click overlay closes', () => {
      const { getByText } = render(
        <DropdownMenu closeOnClickOverlay>
          <DropdownItem label="产品" options={productOptions} />
        </DropdownMenu>,
      );
      fireEvent.click(getByText('产品'));
      act(() => {
        document.body.click();
      });
      expect(getByText('产品')).toBeTruthy();
    });

    it(':change single select', () => {
      let value = 'all';
      const handleChange = vi.fn((v) => {
        value = v;
      });
      const { getByText } = render(
        <DropdownMenu>
          <DropdownItem label="产品" options={productOptions} value={value} onChange={handleChange} />
        </DropdownMenu>,
      );
      const title = getByText('产品');
      const item = title.parentElement as HTMLElement;

      fireEvent.click(title);
      // options are rendered in a dropdown panel (possibly a portal); scope the query to the currently opened panel
      const panels1 = document.querySelectorAll('.t-dropdown-item');
      const panel1 = panels1[panels1.length - 1] as HTMLElement;
      fireEvent.click(within(panel1).getByText('最新产品'));
      expect(handleChange).toHaveBeenCalledWith('new');
      expect(item.classList.contains('t-dropdown-menu__item--active')).toBe(false);
    });

    it(':multiple confirm/reset', () => {
      let value = ['all'];
      const handleChange = vi.fn((v) => {
        value = v;
      });
      const handleConfirm = vi.fn();
      const handleReset = vi.fn();
      const { getByText } = render(
        <DropdownMenu>
          <DropdownItem
            label="多选"
            multiple
            options={productOptions}
            value={value}
            onChange={handleChange}
            onConfirm={handleConfirm}
            onReset={handleReset}
          />
        </DropdownMenu>,
      );
      fireEvent.click(getByText('多选'));
      const panels2 = document.querySelectorAll('.t-dropdown-item');
      const panel2 = panels2[panels2.length - 1] as HTMLElement;
      fireEvent.click(within(panel2).getByText('最新产品'));
      fireEvent.click(within(panel2).getByText('确定'));
      expect(handleConfirm).toHaveBeenCalled();

      // 确认会关闭面板，重新打开后再点击重置
      fireEvent.click(getByText('多选'));
      const panels3 = document.querySelectorAll('.t-dropdown-item');
      const panel3 = panels3[panels3.length - 1] as HTMLElement;
      fireEvent.click(within(panel3).getByText('重置'));
      expect(handleReset).toHaveBeenCalled();
    });

    it(':toggle active class', () => {
      const { getByText } = render(
        <DropdownMenu>
          <DropdownItem label="切换" options={productOptions} />
        </DropdownMenu>,
      );
      const title = getByText('切换');
      const item = title.parentElement as HTMLElement;

      // open
      fireEvent.click(title);
      expect(item.classList.contains('t-dropdown-menu__item--active')).toBe(true);

      // close
      fireEvent.click(title);
      expect(item.classList.contains('t-dropdown-menu__item--active')).toBe(false);
    });

    it(':click disabled option noop', () => {
      const handleChange = vi.fn();
      const { getByText } = render(
        <DropdownMenu>
          <DropdownItem label="含禁用项" options={productOptions} onChange={handleChange} />
        </DropdownMenu>,
      );
      fireEvent.click(getByText('含禁用项'));
      const panels4 = document.querySelectorAll('.t-dropdown-item');
      const panel4 = panels4[panels4.length - 1] as HTMLElement;
      fireEvent.click(within(panel4).getByText('禁用选项'));
      expect(handleChange).not.toHaveBeenCalled();
    });

    it(':collapseMenu ref', () => {
      type DropdownMenuRef = {
        collapseMenu: () => void;
      };

      const ref = React.createRef<DropdownMenuRef>();
      const { getByText } = render(
        <DropdownMenu ref={ref}>
          <DropdownItem label="toCollapse" options={productOptions} />
        </DropdownMenu>,
      );

      fireEvent.click(getByText('toCollapse'));
      const title = getByText('toCollapse');
      const item = title.parentElement as HTMLElement;
      expect(item.classList.contains('t-dropdown-menu__item--active')).toBe(true);

      act(() => {
        ref.current?.collapseMenu();
      });

      expect(item.classList.contains('t-dropdown-menu__item--active')).toBe(false);
    });

    it(':multiple reset/confirm fallback behaviors', () => {
      const { getByText, container } = render(
        <DropdownMenu>
          <DropdownItem label="noReset" multiple options={productOptions} value={['all']} />
        </DropdownMenu>,
      );

      fireEvent.click(getByText('noReset'));
      fireEvent.click(getByText('最新产品'));
      fireEvent.click(getByText('重置'));

      const checked = container.querySelector('.t-checkbox__title--checked');
      expect(checked).toBeTruthy();

      let value = ['all'];
      const handleChange = vi.fn((v) => {
        value = v;
      });
      const { getByText: getByText2 } = render(
        <DropdownMenu>
          <DropdownItem label="noConfirm" multiple options={productOptions} value={value} onChange={handleChange} />
        </DropdownMenu>,
      );

      fireEvent.click(getByText2('noConfirm'));
      const panels5 = document.querySelectorAll('.t-dropdown-item');
      const panel5 = panels5[panels5.length - 1] as HTMLElement;
      fireEvent.click(within(panel5).getByText('最新产品'));
      fireEvent.click(within(panel5).getByText('确定'));

      expect(handleChange).toHaveBeenCalled();
      expect(getByText2('noConfirm')).toBeTruthy();
    });
  });

  describe('slots', () => {
    it(':default', () => {
      const { getByText } = render(
        <DropdownMenu>
          <DropdownItem label="custom" options={productOptions}>
            <div>MyCustomChild</div>
          </DropdownItem>
        </DropdownMenu>,
      );
      fireEvent.click(getByText('custom'));
      expect(getByText('MyCustomChild')).toBeTruthy();
    });
  });
});
