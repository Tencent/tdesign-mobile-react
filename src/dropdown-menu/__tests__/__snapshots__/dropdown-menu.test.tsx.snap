// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropdownMenu > props > :defaultValue 1`] = `
<body
  class=""
>
  <div>
    <div
      class="t-dropdown-menu"
    >
      <div
        class="t-dropdown-menu__item"
      >
        <div
          class="t-dropdown-menu__title"
        >
          最新产品
        </div>
        <svg
          class="t-icon t-icon-caret-down-small t-dropdown-menu__icon"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M12.0001 15L17.5 9.5H6.5L12.0001 15Z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;

exports[`DropdownMenu > props > :direction and zIndex 1`] = `
<div>
  <div
    class="t-dropdown-menu"
  >
    <div
      class="t-dropdown-menu__item t-dropdown-menu__item--active"
    >
      <div
        class="t-dropdown-menu__title"
      >
        产品
      </div>
      <svg
        class="t-icon t-icon-caret-up-small t-dropdown-menu__icon t-dropdown-menu__icon--active"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M12.0001 9L17.5 14.5L6.5 14.5L12.0001 9Z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="t-dropdown-item"
      style="z-index: 9999; bottom: calc(100vh - 0px);"
    >
      <div
        class="t-portal-wrapper"
      >
        <div
          class="t-overlay t-overlay-fade-appear t-overlay-fade-appear-active"
          style="position: absolute; z-index: 1000; animation-duration: 240ms;"
        />
        <div
          class="t-popup t-popup--bottom"
          style="z-index: 9999; animation-fill-mode: forwards; position: absolute; overflow: hidden;"
        >
          <div
            class="t-dropdown-item__content t-popup__content"
          >
            <div
              class="t-dropdown-item__body"
            >
              <div
                class="t-radio-group t-dropdown-item__radio-group"
              >
                <div
                  class="t-radio t-radio--left t-radio--block"
                >
                  <input
                    class="t-radio__original"
                    type="radio"
                    value="all"
                  />
                  <div
                    class="t-radio__icon t-radio__icon--left"
                  >
                    <div
                      class="placeholder"
                    />
                  </div>
                  <div
                    class="t-radio__content"
                  >
                    <span
                      class="t-radio__title"
                      style="-webkit-line-clamp: 3;"
                    >
                      全部产品
                    </span>
                  </div>
                  <div
                    class="t-radio__border t-radio__border--left"
                  />
                </div>
                <div
                  class="t-radio t-radio--left t-radio--block"
                >
                  <input
                    class="t-radio__original"
                    type="radio"
                    value="new"
                  />
                  <div
                    class="t-radio__icon t-radio__icon--left"
                  >
                    <div
                      class="placeholder"
                    />
                  </div>
                  <div
                    class="t-radio__content"
                  >
                    <span
                      class="t-radio__title"
                      style="-webkit-line-clamp: 3;"
                    >
                      最新产品
                    </span>
                  </div>
                  <div
                    class="t-radio__border t-radio__border--left"
                  />
                </div>
                <div
                  class="t-radio t-radio--left t-radio--block"
                >
                  <input
                    class="t-radio__original"
                    type="radio"
                    value="hot"
                  />
                  <div
                    class="t-radio__icon t-radio__icon--left"
                  >
                    <div
                      class="placeholder"
                    />
                  </div>
                  <div
                    class="t-radio__content"
                  >
                    <span
                      class="t-radio__title"
                      style="-webkit-line-clamp: 3;"
                    >
                      最火产品
                    </span>
                  </div>
                  <div
                    class="t-radio__border t-radio__border--left"
                  />
                </div>
                <div
                  class="t-radio t-radio--left t-radio--block"
                >
                  <input
                    class="t-radio__original"
                    disabled=""
                    type="radio"
                    value="disabled"
                  />
                  <div
                    class="t-radio__icon t-radio__icon--left t-radio__icon--disabled"
                  >
                    <div
                      class="placeholder"
                    />
                  </div>
                  <div
                    class="t-radio__content"
                  >
                    <span
                      class="t-radio__title t-radio__title--disabled"
                      style="-webkit-line-clamp: 3;"
                    >
                      禁用选项
                    </span>
                  </div>
                  <div
                    class="t-radio__border t-radio__border--left"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropdownMenu > props > :footer 1`] = `
<body
  class="t-overlay--lock"
>
  <div>
    <div
      class="t-dropdown-menu"
    >
      <div
        class="t-dropdown-menu__item t-dropdown-menu__item--active"
      >
        <div
          class="t-dropdown-menu__title"
        >
          自定义
        </div>
        <svg
          class="t-icon t-icon-caret-down-small t-dropdown-menu__icon t-dropdown-menu__icon--active"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M12.0001 15L17.5 9.5H6.5L12.0001 15Z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        class="t-dropdown-item"
        style="z-index: 11600; top: 0px;"
      >
        <div
          class="t-portal-wrapper"
        >
          <div
            class="t-overlay t-overlay-fade-appear t-overlay-fade-appear-active"
            style="position: absolute; z-index: 1000; animation-duration: 240ms;"
          />
          <div
            class="t-popup t-popup--top"
            style="z-index: 11600; animation-fill-mode: forwards; position: absolute; overflow: hidden;"
          >
            <div
              class="t-dropdown-item__content t-popup__content"
            >
              <div
                class="t-dropdown-item__body"
              >
                <div
                  class="t-radio-group t-dropdown-item__radio-group"
                >
                  <div
                    class="t-radio t-radio--left t-radio--block"
                  >
                    <input
                      class="t-radio__original"
                      type="radio"
                      value="all"
                    />
                    <div
                      class="t-radio__icon t-radio__icon--left"
                    >
                      <div
                        class="placeholder"
                      />
                    </div>
                    <div
                      class="t-radio__content"
                    >
                      <span
                        class="t-radio__title"
                        style="-webkit-line-clamp: 3;"
                      >
                        全部产品
                      </span>
                    </div>
                    <div
                      class="t-radio__border t-radio__border--left"
                    />
                  </div>
                  <div
                    class="t-radio t-radio--left t-radio--block"
                  >
                    <input
                      class="t-radio__original"
                      type="radio"
                      value="new"
                    />
                    <div
                      class="t-radio__icon t-radio__icon--left"
                    >
                      <div
                        class="placeholder"
                      />
                    </div>
                    <div
                      class="t-radio__content"
                    >
                      <span
                        class="t-radio__title"
                        style="-webkit-line-clamp: 3;"
                      >
                        最新产品
                      </span>
                    </div>
                    <div
                      class="t-radio__border t-radio__border--left"
                    />
                  </div>
                  <div
                    class="t-radio t-radio--left t-radio--block"
                  >
                    <input
                      class="t-radio__original"
                      type="radio"
                      value="hot"
                    />
                    <div
                      class="t-radio__icon t-radio__icon--left"
                    >
                      <div
                        class="placeholder"
                      />
                    </div>
                    <div
                      class="t-radio__content"
                    >
                      <span
                        class="t-radio__title"
                        style="-webkit-line-clamp: 3;"
                      >
                        最火产品
                      </span>
                    </div>
                    <div
                      class="t-radio__border t-radio__border--left"
                    />
                  </div>
                  <div
                    class="t-radio t-radio--left t-radio--block"
                  >
                    <input
                      class="t-radio__original"
                      disabled=""
                      type="radio"
                      value="disabled"
                    />
                    <div
                      class="t-radio__icon t-radio__icon--left t-radio__icon--disabled"
                    >
                      <div
                        class="placeholder"
                      />
                    </div>
                    <div
                      class="t-radio__content"
                    >
                      <span
                        class="t-radio__title t-radio__title--disabled"
                        style="-webkit-line-clamp: 3;"
                      >
                        禁用选项
                      </span>
                    </div>
                    <div
                      class="t-radio__border t-radio__border--left"
                    />
                  </div>
                </div>
              </div>
              <div>
                自定义footer
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`DropdownMenu > props > :keys 1`] = `
<body
  class=""
>
  <div>
    <div
      class="t-dropdown-menu"
    >
      <div
        class="t-dropdown-menu__item"
      >
        <div
          class="t-dropdown-menu__title"
        >
          B 选项
        </div>
        <svg
          class="t-icon t-icon-caret-down-small t-dropdown-menu__icon"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M12.0001 15L17.5 9.5H6.5L12.0001 15Z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;

exports[`DropdownMenu > props > :optionsColumns 1`] = `
<body
  class="t-overlay--lock"
>
  <div>
    <div
      class="t-dropdown-menu"
    >
      <div
        class="t-dropdown-menu__item t-dropdown-menu__item--active"
      >
        <div
          class="t-dropdown-menu__title"
        >
          双列
        </div>
        <svg
          class="t-icon t-icon-caret-down-small t-dropdown-menu__icon t-dropdown-menu__icon--active"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M12.0001 15L17.5 9.5H6.5L12.0001 15Z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        class="t-dropdown-item"
        style="z-index: 11600; top: 0px;"
      >
        <div
          class="t-portal-wrapper"
        >
          <div
            class="t-overlay t-overlay-fade-appear t-overlay-fade-appear-active"
            style="position: absolute; z-index: 1000; animation-duration: 240ms;"
          />
          <div
            class="t-popup t-popup--top"
            style="z-index: 11600; animation-fill-mode: forwards; position: absolute; overflow: hidden;"
          >
            <div
              class="t-dropdown-item__content t-popup__content"
            >
              <div
                class="t-dropdown-item__body"
              >
                <div
                  class="t-checkbox-group t-dropdown-item__checkbox-group"
                  style="grid-template-columns: repeat(2, 1fr);"
                >
                  <div
                    class="t-checkbox t-checkbox--left t-checkbox--block t-dropdown-item__checkbox-item t-checkbox--tag"
                  >
                    <div
                      class="t-checkbox__content"
                    >
                      <div
                        class="t-checkbox__title"
                        style="-webkit-line-clamp: 3;"
                      >
                        全部产品
                      </div>
                      <div
                        class="t-checkbox__description"
                        style="-webkit-line-clamp: 5;"
                      />
                    </div>
                  </div>
                  <div
                    class="t-checkbox t-checkbox--left t-checkbox--block t-dropdown-item__checkbox-item t-checkbox--tag"
                  >
                    <div
                      class="t-checkbox__content"
                    >
                      <div
                        class="t-checkbox__title"
                        style="-webkit-line-clamp: 3;"
                      >
                        最新产品
                      </div>
                      <div
                        class="t-checkbox__description"
                        style="-webkit-line-clamp: 5;"
                      />
                    </div>
                  </div>
                  <div
                    class="t-checkbox t-checkbox--left t-checkbox--block t-dropdown-item__checkbox-item t-checkbox--tag"
                  >
                    <div
                      class="t-checkbox__content"
                    >
                      <div
                        class="t-checkbox__title"
                        style="-webkit-line-clamp: 3;"
                      >
                        最火产品
                      </div>
                      <div
                        class="t-checkbox__description"
                        style="-webkit-line-clamp: 5;"
                      />
                    </div>
                  </div>
                  <div
                    class="t-checkbox t-checkbox--left t-checkbox--block t-dropdown-item__checkbox-item t-checkbox--tag"
                  >
                    <div
                      class="t-checkbox__content"
                    >
                      <div
                        class="t-checkbox__title t-checkbox__title--disabled"
                        style="-webkit-line-clamp: 3;"
                      >
                        禁用选项
                      </div>
                      <div
                        class="t-checkbox__description t-checkbox__description--disabled"
                        style="-webkit-line-clamp: 5;"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="t-dropdown-item__footer"
              >
                <button
                  class="t-button t-button--size-medium t-button--base t-button--light t-button--rectangle t-dropdown-item__footer-btn t-dropdown-item__reset-btn"
                  type="button"
                >
                  <span
                    class="t-button__content"
                  >
                     
                    重置
                  </span>
                </button>
                <button
                  class="t-button t-button--size-medium t-button--base t-button--primary t-button--rectangle t-dropdown-item__footer-btn t-dropdown-item__confirm-btn"
                  type="button"
                >
                  <span
                    class="t-button__content"
                  >
                     
                    确定
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`DropdownMenu > props > :style 1`] = `
<div>
  <div
    class="t-dropdown-menu my-menu"
    style="color: rgb(255, 0, 0);"
  >
    <div
      class="t-dropdown-menu__item"
    >
      <div
        class="t-dropdown-menu__title"
      >
        样式测试
      </div>
      <svg
        class="t-icon t-icon-caret-down-small t-dropdown-menu__icon"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M12.0001 15L17.5 9.5H6.5L12.0001 15Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
